### PR TITLE
.Net: Make it possible to get the execution settings for a prompt within an AI Selection Service

### LINKED ISSUE #4209: It's not possible to get the execution settings for a prompt within an AI Selection Service
I have the following prompt function...

```csharp
// Create a template for chat with settings
var chat = kernel.CreateFunctionFromPrompt(
    new PromptTemplateConfig()
    {
        Name = "Chat",
        Description = "Chat with the assistant.",
        Template = @"{{ConversationSummaryPlugin.SummarizeConversation $history}}
        User: {{$request}}
        Assistant: ",
        TemplateFormat = "semantic-kernel",
        InputVariables = [
            new() { Name = "history", Description = "The history of the conversation.", IsRequired = false, Default = "" },
            new() { Name = "request", Description = "The user's request.", IsRequired = true }
        ],
        ExecutionSettings = [
            new() {
                // Default settings for all models
                ExtensionData = new() {
                    { "MaxTokens", 1000 },
                    { "Temperature", 0 }
                }
            },
            new() {
                // Settings for gpt-3.5-turbo
                ModelId = "gpt-3.5-turbo",
                ExtensionData = new() {
                    { "MaxTokens", 4000 },
                    { "Temperature", 0.1 }
                }
            },
            new() {
                // Settings for gpt-4
                ModelId = "gpt-4",
                ExtensionData = new() {
                    { "MaxTokens", 8000 },
                    { "Temperature", 0.3 }
                }
            }
        ]
    }
);
```

This function has execution settings for gpt-3.5-turbo and gpt-4. I now want to build a custom AI selection service that will use this information (e.g., reuse the gpt-4 settings for _any_ version of gpt-4), but it doesn't appear that I have access to this information.

The `TrySelectAIService` has the following method signature

```csharp
public bool TrySelectAIService<T>(
        Kernel kernel, KernelFunction function, KernelArguments arguments,
        [NotNullWhen(true)] out T? service, out PromptExecutionSettings? serviceSettings) where T : class, IAIService
```

I expected to find the execution settings in the `KernelFunction function`, but the execution settings are marked as an internal property.

image

### PR DESCRIPTION
**Pull request by markwallace-microsoft**
### Motivation and Context

Resolves #4209 
### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs [hunk 1]
```diff
@@ -74,7 +74,7 @@ public abstract class KernelFunction
     /// <summary>
     /// Gets the prompt execution settings.
     /// </summary>
-    internal List<PromptExecutionSettings>? ExecutionSettings { get; }
+    public List<PromptExecutionSettings>? ExecutionSettings { get; }
```
Comment (stephentoub) [inline dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs:77]: IReadOnlyList? Otherwise this is mutable.

Comment (stephentoub) [reply dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs:77]: Also, is it OK that the execution settings themselves are mutable via this?

Comment (markwallace-microsoft) [reply dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs:77]: Yeah worth investigating to see can everything we made read only to start with

File: dotnet/src/SemanticKernel.Abstractions/AI/PromptExecutionSettings.cs [hunk 2]
```diff
@@ -30,8 +30,11 @@ public class PromptExecutionSettings
     public string? ModelId { get; set; }
 
     /// <summary>
-    /// Extra properties
+    /// Extra properties that may be included in the serialized execution settings.
     /// </summary>
+    /// <remarks>
+    /// Avoid using this property if possible. Instead, used on of the classes that extends <see cref="PromptExecutionSettings"/>.
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Abstractions/AI/PromptExecutionSettings.cs:36]:
```suggestion
    /// Avoid using this property if possible. Instead, use one of the classes that extends <see cref="PromptExecutionSettings"/>.
```

**Review by dmytrostruk — APPROVED** [review]

**Review by RogerBarreto — APPROVED** [review]

File: dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs (hunk 1 above)
Comment (markwallace-microsoft) [reply dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs]: Created this task to track https://github.com/microsoft/semantic-kernel/issues/4224

**Pull request merged**